### ISSUE TITLE
Traffic shadowing to canary stack

### ISSUE THREAD

**Issue by jessesuen** [argoproj/argo-rollouts#563]
With some ingress controllers such as NGINX and Istio, it is possible to shadow traffic to services. We should leverage these features to allow traffic shadowing to the canary stack during an update.

Proposed syntax:

```yaml
spec:
  strategy:
    canary:
      steps:
      # shadow traffic to canary
      - shadow: {}
      - pause:
           duration: 1h
      - analysis:
          templates:
          - templateName: success
      # disable shadowing and proceed to normal canary
      - shadow:
           disable: true
      - setWeight: 10
      - pause:
           duration: 1h
...
```

**Comment by harikrongali** [comment]
Here is a proposal for the mirror/shadow traffic spec:
Let's call it mirroring is the common name used in istio and other controllers

Two ways mirroring can be used. 
1. Mirroring for canary stack
2. Mirroring for experiment stack


Spec for canary use case:
```
spec:
  strategy:
    canary:
      steps:
     setCanaryScale: {replicas: 2}
      # mirror traffic to canary
      - mirror:
          duration: 1h
          mirrorPercentage: 100.0
         analysis:
            templates:
            - templateName: args-example
            args:
            # required value
            - name: service-name
              value: guestbook-svc.default.svc.cluster.local 
             ....
      - setWeight: 10
      - pause:
           duration: 1h
...

And for the second use case for experiements:
```
...
  strategy:
    canary: 
      steps:
      - experiment:
          duration: 1h
          # mirror traffic
          mirrorPercentage: 100.0%
          templates:
          - name: baseline
            specRef: stable
          - name: canary
            specRef: canary
          analyses:
          - name : mann-whitney
            templateName: mann-whitney
            args:
            - name: baseline-hash
              value: "{{templates.baseline.podTemplateHash}}"
            - name: canary-hash
              value: "{{templates.canary.podTemplateHash}}"
```

**Comment by zachaller** [comment]
Just leaving this note here while I think about this a bit more but with mirroring I could see the next possible request being the need to disable a subset of traffic to the mirror as in you would not want to mirror say traffic that modified data. This can be done a few different ways possibly via path based rules to mirror as in only mirror GET requests etc.

The other option is mirror to a different replica set with say a different configuration such as a patched env var or something that told the service to write to a different database however this dose not work well with the canary replica set and would need to be a new temporary replicaset.

I need to think about this more but just wanted to get something written so I don't forget for now.

**Comment by agrawroh** [comment]
A few thing regarding the proposed specs for the request mirroring [here](https://github.com/argoproj/argo-rollouts/issues/563#issuecomment-1093367015).

1. `mirror` could be named a little bit better. How about `mirroringPolicies` or `trafficMirroringPolicies`? It's okay to use `shadow` instead of `mirror` if the former is more clear.
2. `mirrorPercentage` should just be called `percentage` or `fraction`. `mirror` keyword is redundant here as it's already inferred.
3. Having a `duration` here doesn't make sense. We should either mirror to the entirety or just don't mirror. Basically, how do we answer this question at any point in time: _"Are we mirroring/shadowing traffic?"_
4. Why do we have the entire `analysis` under `mirror`? It's a little confusing as I'd just expect a bunch of upstream services to which we should mirror. Isn't that everything we need?

I think the `specs` should look something like this:
```yaml
...
- trafficMirroringPolicies:
     fraction: 0.2
     destinations:
     - guestbook-svc.default.svc.cluster.local 
     - phonetool-svc.default.svc.cluster.local 
```

I don't think there is anything wrong in supporting multiple destinations to mirror the traffic to.

cc @jessesuen @harikrongali @zachaller What do you guys think?

**Comment by zachaller** [comment]
Yea I actually am planning to add a bit to this proposal because a few things are critical I think to make it truly useful.

Recently we also made a decision to add start and stop steps as the standard for keeping steps simple you can see this with the header based routing. This would remove duration.

I will think about some of the other suggestions more when I re propose this feature.

**Comment by zachaller** [comment]
closed by #2074
